src/hotspot/share/services/mallocHeader.inline.hpp

```diff
@@ -87,11 +77,40 @@ inline bool MallocHeader::check_block_integrity(char* msg, size_t msglen, addres
   // we test the smallest alignment we know.
   // Should we ever start using std::max_align_t, this would be one place to
   // fix up.
-  if (!is_aligned(this, sizeof(uint64_t))) {
-    *p_corruption = (address)this;
+  if (!is_aligned(ptr, sizeof(uint64_t))) {
     jio_snprintf(msg, msglen, "block address is unaligned");
     return false;
   }
+  return true;
+}
+
+template<typename IN, typename OUT>
+inline OUT MallocHeader::resolve_checked_impl(IN memblock){
+  char msg[256];
+  address corruption = NULL;
+  if (!is_valid_malloced_pointer(memblock, msg, sizeof(msg))) {
+    fatal("NMT corruption: Block at " PTR_FORMAT ": %s", memblock, msg);
+  }
+  OUT header_pointer = (OUT)memblock - 1;
+  if (!header_pointer->check_block_integrity( msg, sizeof(msg), &corruption)) {
+    if (corruption != NULL) {
+      MallocHeader::print_block_on_error(tty, (address)memblock);
```

Comment:
Please either pass header pointer explicitly, or just leave the original method non-statiic1 and call it like before via header_pointer.